### ISSUE TITLE
Error out when parameters() is called on replicated models

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#38493]
In DataParallel's implementation, parameters on replicated models are not leaves, and hence are not populated to `_parameters` (see #33907). We should print a clear error message when user code calls `parameters()` on a replicated model.

cc @ngimel

**Comment by choidongyeon** [comment]
Hey I'd be interested in working on this issue. Is this just a matter of adding an assertion error?

**Comment by SsnL** [comment]
This might have BC issue.... Say your model outputs parameter norm as part of the output.

**Comment by mrshenli** [comment]
@SsnL exactly, this has already caused BC issues in existing applications. @soumith mentioned this fix (https://github.com/huggingface/transformers/pull/4300) in huggingface transformers code. 

For context, @soumith @gchanan @ngimel and I did a discussion before posting this issue. We talked about two options:

1. Let `replicate.py` still populate `_parameters` on replicated models to avoid breaking existing apps and find a different fix for LSTM + DP, or
2. Highlight this as a BC-breaking change and communicate it clearly that user code should not try to access `parameters()` on replicated models?

The current consensus is go with option 2, and hence posted this issue to track. Please do comment if you see strong reasons to choose option 1 or if there are better options here. Thanks!

@choidongyeon Thanks for volunteering. 

* I am not sure where the fix should be inserted. We probably don't want to add this to `nn.Module.parameters()` as most use cases are not for replicated models. Another option might be doing this in `replicate.py` by replacing the `parameters()` method on the replicated model object. @ngimel any suggestion here?
* Beside the error message, we will also need to add a warning to `DataParallel`'s doc string.

**Comment by choidongyeon** [comment]
@mrshenli Great, will take a look and get a PR out.

Also, what exactly is "BC issues"?

**Comment by mrshenli** [comment]
> Also, what exactly is "BC issues"? My guess is backwards compatible?

@choidongyeon sorry about the delay. Yes, we were referring to backwards compatibility issues.

**Comment by malfet** [comment]
No activity since 1.5 release, removing milestone label
